AArch64: Enable NUMA for FDT

MrXinWang opened the pull request:
This PR will make `--numa` and related config from Cloud Hypervisor command line work on AArch64 platform with FDT.

MrXinWang commented on docs/memory.md:
```diff
@@ -439,7 +439,9 @@ workload run more efficiently.
-each other with the `:` separator.
+each other with the `:` separator. Note that the memory zone id used by
+different NUMA nodes should also be different, i.e. this is not allowed:
+`--numa guest_numa_id=0,memory_zones=mem0 guest_numa_id=1,memory_zones=mem0`
```
Hi @sboeuf, could you please confirm this change is correct? As from my understanding, different NUMA id cannot share the same memory zone...If this is correct, should we also add some code to forbid user setting command line like this? Thanks!

sboeuf replied:
Yes you're totally right. We can't have the same memory zone belonging to more than one NUMA node. If this is not validated in the code, then yes we should add a check to prevent the user from specifying such configuration.

sboeuf replied:
I'd rephrase with:
```
Note that a memory zone must belong to a single NUMA node. The following
configuration is incorrect, therefore not allowed:
`--numa guest_numa_id=0,memory_zones=mem0 guest_numa_id=1,memory_zones=mem0`
```

MrXinWang replied:
> Yes you're totally right. We can't have the same memory zone belonging to more than one NUMA node. If this is not validated in the code, then yes we should add a check to prevent the user from specifying such configuration.

AFAIK the current code is not validating that - As I remember I tried this once...

Then I will add some checking code. How about this? @sboeuf

Here (https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/vmm/src/config.rs#L1689) when we parsing the config, we add a global/static data structure, for example a vector, to store the already used memory zones. If we meet a used memory zone when parsing the rest of NUMA nodes, we raise an error.

michael2012z commented on vmm/src/vm.rs:
```diff
@@ -651,7 +646,6 @@ impl Vm {
         })
     }
 
-    #[cfg(feature = "acpi")]
```
@rbradford , is there any problem on X86 if we make NUMA not specific to ACPI?

rbradford replied:
Should be fine, the `--numa` parsing is not feature gated.

rbradford reviewed (approved):
You have a rubber stamp from me but I think @sboeuf should answer the question about zones since he is the expert there.

MrXinWang commented:
> You have a rubber stamp from me but I think @sboeuf should answer the question about zones since he is the expert there.

of course :))

sboeuf commented on vmm/src/vm.rs:
```diff
@@ -519,7 +518,7 @@ pub struct Vm {
     #[cfg(all(feature = "kvm", target_arch = "x86_64"))]
     saved_clock: Option<hypervisor::ClockData>,
-    #[cfg(feature = "acpi")]
+    #[allow(dead_code)]
```
Is it the way we want to handle this? The alternative would be to provide the features for which this is relevant.

MrXinWang replied:
Yeah (more or less), as the clippy will also check `--features=kvm` without `acpi` on x86_64, and in this case this `numa_nodes: NumaNodes` field is never used. The `#[allow(dead_code)]` is intended just to mute clippy.

On arm everything is fine (both `--features=kvm` and `--features=kvm, acpi` will use this field).

It is kinda tricky to provide a proper gate here...

sboeuf commented on tests/integration.rs:
```diff
@@ -2337,6 +2337,63 @@ mod tests {
 
+        #[test]
+        #[cfg(target_arch = "aarch64")]
+        fn test_aarch64_guest_numa_nodes_dt() {
```
Why do you need a dedicated test?
You should be able to reuse `test_guest_numa_nodes()`.
I think you should reuse `test_guest_numa_nodes()` to make it a common function that is called from `test_guest_numa_nodes_acpi()` and `test_guest_numa_nodes_dt()`. Of course you would apply the correct feature gating on top of each to make sure they can run as expected.

MrXinWang replied:
Ah, this is because we want to make sure that both the DT and ACPI will be tested on Arm. For now, we only build the Cloud Hypervisor for once in CI using `cargo build`. Using feature gates cannot cover both cases.

But I like the idea of reusing `test_guest_numa_nodes()` as a common function. Let me do some refactoring :)) Thank you very much for the suggestion!

rbradford replied:
Can you can test the two code paths by firmware boot vs direct boot (acpi vs fdt)?

MrXinWang replied:
> Can you can test the two code paths by firmware boot vs direct boot (acpi vs fdt)?

Hi @rbradford ! Not very sure what you mean (as there is a typo), but yeah, the idea is just to change the path of `--kernel`. However on arm we need to run both of two cases, i.e. `--kernel=direct_kernel_boot_path()` once and `--kernel=edk2_path()` once. I don't think this is possible in one single test case unless using a loop. So I would extract most of the common code into a helper function and split the original test case into `test_guest_numa_nodes_acpi()` and `test_guest_numa_nodes_dt()`

rbradford replied:
This is exactly what I was thinking.

MrXinWang commented:
Temporary making this PR as draft to fix some useful comments.

MrXinWang commented:
Hi @sboeuf @rbradford ! I believe I have addressed all of your concerns:

- https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2950/commits/2e925d6c67f8fb3bbeafbca927c195401b47f42a and https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2950/commits/27401b1765688b41b751e70b4ed522f251698cee add proper gates to the NUMA related data structures and functions.
- https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2950/commits/2dc1ed271d11b3d9629123e1ddd6db7ee75a6f98 adds validation code to `--numa` memory nodes, an example is given below if `mem1` is used twice:
```
Error parsing config: Error validating configuration: Memory zone: mem1 belongs to multiple NUMA nodes
```
- https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2950/commits/240ac48ad58690b036630463afd985e6a58e5ef0 extracts most of the common code into a helper function and split the original test case into `test_guest_numa_nodes_acpi()` and `test_guest_numa_nodes_dt()`

rbradford commented on arch/src/lib.rs:
```diff
@@ -96,6 +101,46 @@ fn pagesize() -> usize {
+    pub sgx_epc_sections: Vec<SgxEpcSection>,
+}
+
+impl NumaNode {
```
What's the benefit of having all these accessors if all the fields are pub?

MrXinWang replied:
To be honest I don't know....I didn't think too much when moving this part and just moved the original code as it is.

Maybe @sboeuf is more experienced to answer this question?

sboeuf replied:
Check if the code is doing extensive use of directly accessing the fields. If not, then try to make sure it relies on accessors only.

rbradford replied:
I think you can remove the accessors if all the fields are pub.

MrXinWang replied:
Hi @rbradford. Yes and done. Such change is reflected in https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2950/commits/fdd0a6c469b638ee5c79ee059a71bb9d2f1c578c

rbradford commented on vmm/src/config.rs:
```diff
@@ -1915,6 +1920,20 @@ impl VmConfig {
+                    if !used_numa_node_memory_zones.contains(memory_zone) {
+                        used_numa_node_memory_zones.push(memory_zone.to_string());
+                    } else {
+                        return Err(ValidationError::MemoryZoneReused(memory_zone.to_string()));
```
This is nice, but could your also give the conficting node ids to make it easier for the user to find their conflict?

MrXinWang replied:
Done. Yes that makes even more sense :))

sboeuf reviewed (commented):
Integration test looks good, thanks for fixing it!

sboeuf reviewed (approved):
Great! Thanks for addressing all the comments and issues!
